<img src="https://upload.wikimedia.org/wikipedia/commons/7/76/Grain_logo.png" style="float: left; margin: 0px; height: 80px">

<div style="clear: both;"></div>

<div>
    <h1>Capstone project: Business dashboard</h1>
    <p>Part 1 - Data cleaning</p>
</div>

---
## Problem Statement
The F&B industry in Singapore is valued at $13 billion a year. The food catering industry specifically is picking up post-covid19 and is expected to grow in the next few years. However, there are many issues that catering companies have to solve. We will focus on 3 issues.

1. Inventory management (Given past data, can we forecast demand to efficiently carry out inventory management?)
2. Price negotiation (What is the recommended price for customers from different sectors?)
3. Food quality: Delivery driver allocation (Is there a way to automatically assign drivers for deliveries to pick up hot food at the right timings and to effectively distribute the deliveries so that only a final check is needed?)

## Contents:
- [Import libraries](#Import-libraries) 
- [Import data](#Import-data)
- [Data cleaning](#Data-cleaning)
- [Data imputation](#Data-imputation)

# Import libraries 

In [1]:
import pandas as pd

# Import data 

In [2]:
# Import data from data file
# This is 1 year worth of orders for Minibuffet
df = pd.read_csv("../data/Mini_buffets_2022.csv")

# Set the view limit to none
pd.set_option('display.max_columns', None) 
print(df.shape)
df.head()

(11475, 40)


event_id     order_id event_state sales_order_state sales_type  trial  \
0    E8162  SO070833480   PUBLISHED         CONFIRMED    inbound  False   
1    F1912  SO120836198   PUBLISHED         CONFIRMED   outbound  False   
2    F1914  SO120836198   PUBLISHED         CONFIRMED   outbound  False   
3    F2598  SO121236841   PUBLISHED         CONFIRMED    inbound  False   
4    F2672  SO121236916   PUBLISHED         CONFIRMED     repeat  False   

  promo_code referred_by     sector                              company_name  \
0        NaN         NaN        NaN                                       NaN   
1        NaN         NaN  Hospitals  National University Hospital (S) Pte Ltd   
2        NaN         NaN  Hospitals  National University Hospital (S) Pte Ltd   
3   RF991822         NaN        NaN                                       NaN   
4        NaN         NaN        NaN                                       NaN   

   customer_id        type  sub_type  cost_per_pax  lines  ways  pax  \
0        45602  MiniBuffet       NaN         18.00      1     1   15   
1        31632  MiniBuffet       NaN         12.99      1     1  160   
2        31632  MiniBuffet       NaN         12.99      1     1   90   
3        49671  MiniBuffet       NaN         33.60      1     1    5   
4        47363  MiniBuffet       NaN         18.00      1     1  126   

                                             address  postal_code  \
0              120 King's Rd, Church of St. Ignatius       268172   
1      1E Kent Ridge Road, NUHS Tower Block, Level 8       119228   
2      1E Kent Ridge Road, NUHS Tower Block, Level 8       119228   
3                        123 Teck Whye Lane, #06-850       680123   
4  Amazing Grace Presbyterian Church, 408B Upper ...       466484   

  need_to_use_stairs                               special_instructions  \
0                NaN  Customised label: Thank you for making this sp...   
1                NaN  160 bento & cutlery | 160 fruit parcels | 250 ...   
2                NaN              90 bento & cutlery | 90 fruit parcels   
3                NaN  Please hand over the delivery to a household m...   
4                NaN  252 x Bento + Cutlery \n\n130 x paper bags\n\n...   

   need_pre_setup pre_setup_date_time                  created_at  \
0           False                 NaN   2021-07-08T12:25:35.03374   
1           False                 NaN  2021-12-08T21:50:13.615565   
2           False                 NaN  2021-12-08T22:02:35.998535   
3           False                 NaN  2021-12-12T10:25:59.331641   
4           False                 NaN  2021-12-12T23:30:49.300378   

                   updated_at     pickup_date_time      setup_date_time  \
0  2022-02-24T16:17:14.542858  2022-02-26T10:45:00  2022-02-26T11:45:00   
1  2022-02-21T17:15:50.026876  2022-02-23T09:30:00  2022-02-23T10:30:00   
2  2022-02-21T17:15:51.700405  2022-02-23T09:30:00  2022-02-23T10:30:00   
3  2022-01-06T13:55:04.219927  2022-01-08T18:00:00  2022-01-08T19:00:00   
4  2022-01-06T13:55:10.589043  2022-01-08T10:30:00  2022-01-08T11:30:00   

     serving_date_time   takedown_date_time      ready_to_eat_at  \
0  2022-02-26T11:45:00  2022-02-26T11:45:00  2022-02-26T10:45:00   
1  2022-02-23T10:30:00  2022-02-23T10:30:00  2022-02-23T09:30:00   
2  2022-02-23T10:30:00  2022-02-23T10:30:00  2022-02-23T09:30:00   
3  2022-01-08T19:00:00  2022-01-08T19:00:00  2022-01-08T18:00:00   
4  2022-01-08T11:30:00  2022-01-08T11:30:00  2022-01-08T10:30:00   

            consume_by  sub_total_value  additional_charges_value  \
0  2022-02-26T14:45:00            270.0                     77.85   
1  2022-02-23T13:30:00           2078.4                      0.00   
2  2022-02-23T13:30:00           1169.1                      0.00   
3  2022-01-08T22:00:00            168.0                     44.81   
4  2022-01-08T14:30:00           2268.0                     65.00   

   delivery_fee  surcharge_value  discount_value  revenue  \
0          30.

# Data cleaning

**Remove the rows where the event_state is DELETED as a customer may have placed an order but later changed their mind and cancelled it**

In [3]:
# Filter out rows where 'event_state' is not equal to 'DELETED'
df = df[df['event_state'] != 'DELETED']
df.shape

(11121, 40)

There were initially 11, 475 rows and after removing the deleted events, we are left with 11, 121 rows. This means that there were 354 deleted events.

**Choose the columns that we want**

In [4]:
# Check if the event ID is unique by adding up the sum of all the unique IDs
num_of_events = df['event_id'].nunique()
num_of_events

11121

Yes the event ID's are unique since the number of unique event ID's is equal to the number of rows in the dataframe

In [5]:
# Reorganise the dataframe
new_df = df[['event_id', 'event_state', 'sales_type', 'sector', 'company_name', 'customer_id', 'cost_per_pax', 'pax', 'address', 'postal_code', 'special_instructions', 'pickup_date_time', 'serving_date_time', 'revenue']]
print(new_df.shape)
new_df.head(60)

(11121, 14)


event_id event_state sales_type                                 sector  \
0     E8162   PUBLISHED    inbound                                    NaN   
1     F1912   PUBLISHED   outbound                              Hospitals   
2     F1914   PUBLISHED   outbound                              Hospitals   
3     F2598   PUBLISHED    inbound                                    NaN   
4     F2672   PUBLISHED     repeat                                    NaN   
5     F2830   PUBLISHED     repeat        Oil & Gas, Chemicals and Energy   
6     F3176   PUBLISHED    inbound                                    NaN   
7     F3177   PUBLISHED    inbound                                    NaN   
8     F3201   PUBLISHED     repeat                         Pharmaceutical   
9     F3202   PUBLISHED     repeat                         Pharmaceutical   
10    F3207   PUBLISHED    inbound                              Education   
11    F3213   PUBLISHED    inbound  Government, Nonprofits & Associations   
13    F3215   PUBLISHED    inbound                                    NaN   
14    F3216   PUBLISHED     repeat                         Pharmaceutical   
15    F3217   PUBLISHED    inbound                                    NaN   
16    F3228   PUBLISHED    inbound                                    NaN   
18    F3230   PUBLISHED     repeat                                    NaN   
20    F3233   PUBLISHED     repeat                                    NaN   
22    F3235   PUBLISHED    inbound                              Hospitals   
23    F3236   PUBLISHED    inbound          Manufacturing and Engineering   
24    F3239   PUBLISHED     repeat                              Hospitals   
25    F3240   PUBLISHED     repeat                              Hospitals   
26    F3241   PUBLISHED     repeat                              Hospitals   
27    F3242   PUBLISHED     repeat                              Hospitals   
28    F3243   PUBLISHED   outbound                              Hospitals   
30    F3245   PUBLISHED     repeat                              Education   
31    F3246   PUBLISHED    inbound                                    NaN   
32    F3247   PUBLISHED   outbound                              Hospitals   
33    F3248   PUBLISHED     repeat                                    NaN   
34    F3251   PUBLISHED    inbound                                    NaN   
35    F3252   PUBLISHED   outbound                              Education   
36    F3253   PUBLISHED    inbound                  Technology (software)   
37    F3254   PUBLISHED    inbound                  Technology (software)   
38    F3256   PUBLISHED     repeat                                    NaN   
39    F3257   PUBLISHED    inbound                              Hospitals   
40    F3258   PUBLISHED    inbound                         Events Company   
41    F3259   PUBLISHED    inbound  Government, Nonprofits & Associations   
42    F3260   PUBLISHED    inbound  Government, Nonprofits & Associations   
43    F3261   PUBLISHED    inbound                                    NaN   
44    F3262   PUBLISHED    inbound                                    NaN   
45    F3263   PUBLISHED    inbound                                    NaN   
46    F3265   PUBLISHED    inbound                                    NaN   
47    F3266   PUBLISHED    inbound                                    NaN   
48    F3267   PUBLISHED     repeat                                    NaN   
49    F3268   PUBLISHED    inbound                                    NaN   
50    F3269   PUBLISHED    inbound                                    NaN   
51    F3270   PUBLISHED     repeat                         Events Company   
52    F3271   PUBLISHED     repeat                         Events Company   
53    F3272   PUBLISHED     repeat                         Events Company   
54    F3274   PUBLISHED     repeat                         Events Company   
55    F3275   PUBLISHED     repeat                         Events Company   

We chose 14 relevant columns from the 40 that were in the dataset.

**Check for duplicate rows**

In [6]:
# Assuming your DataFrame is called 'df'
num_duplicate_rows = new_df.duplicated().sum()
num_duplicate_rows

0

There are no duplicated rows. 

**Check for null values**

In [7]:
# Get the data types by column
data_types_by_column = new_df.dtypes

# Get the number of null values by column
null_values_by_column = new_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((new_df.isnull().sum() / len(new_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object         4453                          40.0
company_name            object         2596                          23.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object         1502                          14.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

There are quite a lot of missing values for the sector, company_name and special_instructions. We will impute the missing values in the next segement.

# Data imputation

**Impute data for sector column**

First we check if there are rows where the sector is missing but the company name is not missing. Then we will search the dataframe for rows with the same company name with the sector value. We will then put in the sector values for those rows where the sector is missing. 

In [8]:
# Visualise which company has the most missing values. 
# Group by 'company_name' and count missing 'sector' values
missing_sector_counts = new_df.groupby('company_name')['sector'].apply(lambda x: x.isnull().sum()).reset_index(name='missing_sector_count')

# Sort in descending order based on the count of missing values
missing_sector_counts = missing_sector_counts.sort_values(by='missing_sector_count', ascending=False)

# Remove rows where the number of missing_sector_count is 0
missing_sector_counts = missing_sector_counts[missing_sector_counts['missing_sector_count'] > 0]
missing_sector_counts.head(12)

company_name  missing_sector_count
263                                           CaterSpot                   172
1305  Singapore University of Technology and Design ...                    86
264                                          Catersmith                    83
329                                       Compass Group                    54
674   Institute of Singapore Chartered Accountants (...                    43
93                                               Amazon                    38
146                              Aureus Group Pte. Ltd.                    36
1498                                 VISIONNAIRE EVENTS                    32
512                                            FoodLine                    24
330                    Compass Group Singapore Pte Ltd.                    23
576                                               Grain                    17
1517                                                WPP                    13

**Impute sector based on campany name and ID**

In [9]:
filtered_df = new_df[new_df['company_name'].str.contains('Institute of Singapore Chartered Accountant', case=False, na=False)]
print(len(filtered_df))
filtered_df

52


event_id event_state sales_type                 sector  \
5857     G1033   PUBLISHED   outbound                    NaN   
5858     G1034   PUBLISHED   outbound                    NaN   
6050     G1359   PUBLISHED     repeat                    NaN   
6079     G1409   PUBLISHED     repeat                    NaN   
6756     G2654   PUBLISHED     repeat                    NaN   
7335     G3865   PUBLISHED     repeat                    NaN   
7362     G3971   PUBLISHED     repeat                    NaN   
7416     G4102   PUBLISHED     repeat                    NaN   
7453     G4195   PUBLISHED     repeat                    NaN   
7532     G4392   PUBLISHED     repeat                    NaN   
7533     G4393   PUBLISHED     repeat                    NaN   
7534     G4395   PUBLISHED     repeat                    NaN   
7591     G4489   PUBLISHED     repeat                    NaN   
7634     G4566   PUBLISHED     repeat                    NaN   
7635     G4567   PUBLISHED     repeat                    NaN   
7673     G4686   PUBLISHED     repeat                    NaN   
7717     G4779   PUBLISHED     repeat                    NaN   
7720     G4790   PUBLISHED     repeat                    NaN   
7721     G4791   PUBLISHED     repeat                    NaN   
7815     G4994   PUBLISHED     repeat                    NaN   
7931     G5201   PUBLISHED     repeat                    NaN   
7932     G5204   PUBLISHED     repeat                    NaN   
7933     G5205   PUBLISHED     repeat                    NaN   
7934     G5206   PUBLISHED     repeat                    NaN   
8239     G5764   PUBLISHED     repeat                    NaN   
8240     G5765   PUBLISHED     repeat                    NaN   
8242     G5768   PUBLISHED     repeat                    NaN   
8390     G6095   PUBLISHED     repeat                    NaN   
8392     G6097   PUBLISHED     repeat                    NaN   
8394     G6099   PUBLISHED     repeat                    NaN   
8607     G6501   PUBLISHED     repeat                    NaN   
8609     G6503   PUBLISHED     repeat                    NaN   
8610     G6504   PUBLISHED     repeat                    NaN   
8962     G7119   PUBLISHED     repeat                    NaN   
8963     G7120   PUBLISHED     repeat                    NaN   
8966     G7123   PUBLISHED     repeat                    NaN   
9205     G7551   PUBLISHED     repeat                    NaN   
9398     G7892   PUBLISHED     repeat                    NaN   
9400     G7895   PUBLISHED     repeat                    NaN   
9695     G8383   PUBLISHED     repeat                    NaN   
9858     G8653   PUBLISHED   outbound                    NaN   
9952     G8791   PUBLISHED     repeat                    NaN   
9953     G8792   PUBLISHED     repeat                    NaN   
10243    G9237   PUBLISHED     repeat  Professional services   
10280    G9302   PUBLISHED     repeat  Professional services   
10281    G9303   PUBLISHED     repeat  Professional services   
10323    G9407   PUBLISHED     repeat  Professional services   
10324    G9408   PUBLISHED     repeat  Professional services   
10893    H0542   PUBLISHED     repeat  Professional services   
10894    H0543   PUBLISHED     repeat  Professional services   
10895    H0544   PUBLISHED     repeat  Professional services   
10962    H0734   PUBLISHED   outbound  Professional services   

                                            company_name  customer_id  \
5857   Institute of Singapore Chartered Accountants (...        49733   
5858   Institute of Singapore Chartered Accountants (...        49733   
6050   Institute of Singapore Chartered Accountants (...        49733   
6079   Institute of Singapore Chartered Accountants (...        49733   
6756   Institute of Singapore Chartered Accountants (...        49733   
7335   Institute of Singapore Chartered Accountants (...        49733   
7362   Institute of Singapore Chartered Accountants (...        49733   
7416   Institute of Singapore

From this we can see that there are some instances where the company_name is not null and there are some instances where the sector is filled up and other instances where the sector is a null value. We will impute the null values with the sector values from the other rows where the sector is not null. 

In [10]:
# Create a separate group for rows where 'company_name' is null
null_company_group = new_df[new_df['company_name'].isnull()]

# Group the DataFrame by 'company_name' (excluding null values)
groups = new_df[new_df['company_name'].notnull()].groupby('company_name', group_keys=True)

# Define a function to impute the sector within each group
def impute_sector(group):
    non_null_sectors = group['sector'].dropna()
    
    if not non_null_sectors.empty:
        first_non_null_sector = non_null_sectors.iloc[0]
        group['sector'].fillna(first_non_null_sector, inplace=True)
    
    return group

# Apply the impute_sector function to each group
imputed_groups = groups.apply(impute_sector)

# Concatenate the null_company_group and imputed_groups back together
imputed_df = pd.concat([null_company_group, imputed_groups])

# Reset the index of the imputed DataFrame
imputed_df.reset_index(drop=True, inplace=True)

# Print the resulting DataFrame
imputed_df

event_id event_state sales_type sector     company_name  customer_id  \
0        E8162   PUBLISHED    inbound    NaN              NaN        45602   
1        F2598   PUBLISHED    inbound    NaN              NaN        49671   
2        F2672   PUBLISHED     repeat    NaN              NaN        47363   
3        F3176   PUBLISHED    inbound    NaN              NaN        50203   
4        F3177   PUBLISHED    inbound    NaN              NaN        46587   
...        ...         ...        ...    ...              ...          ...   
11116    F4360   PUBLISHED     repeat    NaN  wyeth nutrition        16441   
11117    F4769   PUBLISHED     repeat    NaN  wyeth nutrition        24879   
11118    F4771   PUBLISHED     repeat    NaN  wyeth nutrition        24879   
11119    G9673   PUBLISHED     repeat    NaN  wyeth nutrition        24879   
11120    F8779   PUBLISHED    inbound    NaN        yposg.org        54878   

       cost_per_pax  pax                                            address  \
0             18.00   15              120 King's Rd, Church of St. Ignatius   
1             33.60    5                        123 Teck Whye Lane, #06-850   
2             18.00  126  Amazing Grace Presbyterian Church, 408B Upper ...   
3              0.00   10                      910 Hougang Street 91, #06-80   
4             16.73   18                   Block 432 Clementi Ave 3 #06-294   
...             ...  ...                                                ...   
11116         40.91    5       Blk 2 Balestier Hill Shopping Centre #04-669   
11117         33.60    5  Mt alvernia Hospital  820 Thomson Road #07-63 ...   
11118         30.00   10  Mt alvernia Hospital  820 Thomson Road #07-63 ...   
11119          0.00   40  100 Bukit Timah Road, KKH Hospital, Dietetics ...   
11120         22.50   16                          64-1 Spottiswoode Park Rd   

       postal_code                               special_instructions  \
0           268172  Customised label: Thank you for making this sp...   
1           680123  Please hand over the delivery to a household m...   
2           466484  252 x Bento + Cutlery \n\n130 x paper bags\n\n...   
3           530910  Morning Bundle B and E\n23 x Bentos and Cutleries   
4           120432  Please take the single lift lobby (beside Poon...   
...            ...                                                ...   
11116       320002                                                NaN   
11117       574623                                                NaN   
11118       574623                         deliver togther with F4769   
11119       229899                               40 x bento + cutlery   
11120        88653                                                NaN   

          pickup_date_time    serving_date_time  revenue  
0      2022-02-26T10:45:00  2022-02-26T11:45:00   377.85  
1      2022-01-08T18:00:00  2022-01-08T19:00:00   230.67  
2      2022-01-08T10:30:00  2022-01-08T11:30:00  2333.00  
3      2022-01-08T07:30:00  2022-01-08T08:30:00   428.85  
4      2022-01-08T12:00:00  2022-01-08T13:00:00   321.79  
...                    ...                  ...      ...  
11116  2022-02-03T11:00:00  2022-02-03T12:00:00   229.55  
11117  2022-02-16T11:15:00  2022-02-16T12:15:00   183.00  
11118  2022-02-16T11:15:00  2022-02-16T12:15:00   383.04  
11119  2023-01-05T11:00:00  2023-01-05T12:00:00   579.40  
11120  2022-06-22T17:00:00  2022-06-22T18:00:00   446.00  

[11121 rows x 14 columns]

In [11]:
filtered_df = imputed_df[imputed_df['company_name'].str.contains('Institute of Singapore Chartered Accountant', case=False, na=False)]
print(len(filtered_df))
filtered_df

52


event_id event_state sales_type                 sector  \
6462    G1033   PUBLISHED   outbound  Professional services   
6463    G1034   PUBLISHED   outbound  Professional services   
6464    G1359   PUBLISHED     repeat  Professional services   
6465    G1409   PUBLISHED     repeat  Professional services   
6466    G2654   PUBLISHED     repeat  Professional services   
6467    G3865   PUBLISHED     repeat  Professional services   
6468    G3971   PUBLISHED     repeat  Professional services   
6469    G4102   PUBLISHED     repeat  Professional services   
6470    G4195   PUBLISHED     repeat  Professional services   
6471    G4392   PUBLISHED     repeat  Professional services   
6472    G4393   PUBLISHED     repeat  Professional services   
6473    G4395   PUBLISHED     repeat  Professional services   
6474    G4489   PUBLISHED     repeat  Professional services   
6475    G4566   PUBLISHED     repeat  Professional services   
6476    G4567   PUBLISHED     repeat  Professional services   
6477    G4686   PUBLISHED     repeat  Professional services   
6478    G4779   PUBLISHED     repeat  Professional services   
6479    G4790   PUBLISHED     repeat  Professional services   
6480    G4791   PUBLISHED     repeat  Professional services   
6481    G4994   PUBLISHED     repeat  Professional services   
6482    G5201   PUBLISHED     repeat  Professional services   
6483    G5204   PUBLISHED     repeat  Professional services   
6484    G5205   PUBLISHED     repeat  Professional services   
6485    G5206   PUBLISHED     repeat  Professional services   
6486    G5764   PUBLISHED     repeat  Professional services   
6487    G5765   PUBLISHED     repeat  Professional services   
6488    G5768   PUBLISHED     repeat  Professional services   
6489    G6095   PUBLISHED     repeat  Professional services   
6490    G6097   PUBLISHED     repeat  Professional services   
6491    G6099   PUBLISHED     repeat  Professional services   
6492    G6501   PUBLISHED     repeat  Professional services   
6493    G6503   PUBLISHED     repeat  Professional services   
6494    G6504   PUBLISHED     repeat  Professional services   
6495    G7119   PUBLISHED     repeat  Professional services   
6496    G7120   PUBLISHED     repeat  Professional services   
6497    G7123   PUBLISHED     repeat  Professional services   
6498    G7551   PUBLISHED     repeat  Professional services   
6499    G7892   PUBLISHED     repeat  Professional services   
6500    G7895   PUBLISHED     repeat  Professional services   
6501    G8383   PUBLISHED     repeat  Professional services   
6502    G8653   PUBLISHED   outbound  Professional services   
6503    G8791   PUBLISHED     repeat  Professional services   
6504    G8792   PUBLISHED     repeat  Professional services   
6505    G9237   PUBLISHED     repeat  Professional services   
6506    G9302   PUBLISHED     repeat  Professional services   
6507    G9303   PUBLISHED     repeat  Professional services   
6508    G9407   PUBLISHED     repeat  Professional services   
6509    G9408   PUBLISHED     repeat  Professional services   
6510    H0542   PUBLISHED     repeat  Professional services   
6511    H0543   PUBLISHED     repeat  Professional services   
6512    H0544   PUBLISHED     repeat  Professional services   
6513    H0734   PUBLISHED   outbound  Professional services   

                                           company_name  customer_id  \
6462  Institute of Singapore Chartered Accountants (...        49733   
6463  Institute of Singapore Chartered Accountants (...        49733   
6464  Institute of Singapore Chartered Accountants (...        49733   
6465  Institute of Singapore Chartered Accountants (...        49733   
6466  Institute of Singapore Chartered Accountants (...        49733   
6467  Institute of Singapore Chartered Accountants (...        49733   
6468  Institute of Singapore Chartered Accountants (...        49733   
6469  Institute of Singapore Chartered Accountants (...        49733   
6470  Institute o

From this specific exmaple, we see that the rows with the company name as 'Institute of Singapore Chartered Accountants (ISCA)' and have null values for the sector had their sectors filled up with 'Professional services'.

In [12]:
# Get the data types by column
data_types_by_column = imputed_df.dtypes

# Get the number of null values by column
null_values_by_column = imputed_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((imputed_df.isnull().sum() / len(imputed_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object         4284                          39.0
company_name            object         2596                          23.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object         1502                          14.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

Previously, we had 4453 missing values for the sector column and now we have 4284 missing values in the sector column. We managed to impute 169 values. It is not a lot, so we will impute some of the data manually for the top 10 companies which have missing values. 

**Manually impute the sector for the top 10 missing sectors but have company names**

In [13]:
# Find the row where company_name is 'CaterSpot' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'CaterSpot', 'sector'] = 'Catering service'

# Find the row where company_name is 'Singapore University of Technology and Design (SUTD)' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Singapore University of Technology and Design (SUTD)', 'sector'] = 'Education'

# Find the row where company_name is 'Catersmith' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Catersmith', 'sector'] = 'Catering service'

# Find the row where company_name is 'Compass Group' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Compass Group', 'sector'] = 'Catering service'

# Find the row where company_name is 'Amazon' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Amazon', 'sector'] = 'Technology (software)'

# Find the row where company_name is 'Aureus Group Pte. Ltd.' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Aureus Group Pte. Ltd.', 'sector'] = 'Music school'

# Find the row where company_name is 'VISIONNAIRE EVENTS' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'VISIONNAIRE EVENTS', 'sector'] = 'Events Company'

# Find the row where company_name is 'FoodLine' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'FoodLine', 'sector'] = 'Catering service'

# Find the row where company_name is 'Compass Group Singapore Pte Ltd.' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Compass Group Singapore Pte Ltd.', 'sector'] = 'Catering service'

# Find the row where company_name is 'Grain' and set the sector value
imputed_df.loc[imputed_df['company_name'] == 'Grain', 'sector'] = 'Catering company'


In [14]:
filtered_df = imputed_df[imputed_df['company_name'].str.contains('Grain', case=False, na=False)]
filtered_df.head()

event_id event_state sales_type            sector company_name  \
6172    F3714   PUBLISHED     repeat  Catering company        Grain   
6173    F3825   PUBLISHED     repeat  Catering company        Grain   
6174    F6237   PUBLISHED     repeat  Catering company        Grain   
6175    F6262   PUBLISHED     repeat  Catering company        Grain   
6176    F9677   PUBLISHED     repeat  Catering company        Grain   

      customer_id  cost_per_pax  pax  \
6172        18528         39.04    5   
6173         2558         39.04    5   
6174         2558         10.50   16   
6175         2558         16.00   10   
6176         2558         18.50   10   

                                                address  postal_code  \
6172  5 Burn Road, #05-01 Tee Yih Jia Building,, #05-01       369972   
6173           5 Burn Road, #05-01 Tee Yih Jia Building       369972   
6174           5 Burn Road, #05-01 Tee Yih Jia Building       369972   
6175           5 Burn Road, #05-01 Tee Yih Jia Building       369972   
6176           5 Burn Road, #05-01 Tee Yih Jia Building       369972   

           special_instructions     pickup_date_time    serving_date_time  \
6172                        NaN  2022-01-18T11:00:00  2022-01-18T11:00:00   
6173                        NaN  2022-01-31T17:45:00  2022-01-31T18:45:00   
6174                        NaN  2022-04-13T12:30:00  2022-04-13T13:30:00   
6175  Food tasting + photoshoot  2022-04-14T12:30:00  2022-04-14T13:30:00   
6176                        NaN  2022-07-09T11:00:00  2022-07-09T12:00:00   

      revenue  
6172     0.00  
6173   175.68  
6174     0.00  
6175     0.00  
6176   166.50

In [15]:
# Get the data types by column
data_types_by_column = imputed_df.dtypes

# Get the number of null values by column
null_values_by_column = imputed_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((imputed_df.isnull().sum() / len(imputed_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object         3719                          33.0
company_name            object         2596                          23.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object         1502                          14.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

Previously, we had 4284 missing values for the sector column and now we have 3719 missing values in the sector column. We managed to impute 565 values. We stil have a massive number of missing values so we will impute the sector as missing. 

**Lastly, we will impute 'missing' for the rest of the missing values in the sector column**

In [16]:
# Impute 'sector' as 'missing' for rows with null sector values
imputed_df['sector'].fillna('missing', inplace=True)

In [17]:
# Get the data types by column
data_types_by_column = imputed_df.dtypes

# Get the number of null values by column
null_values_by_column = imputed_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((imputed_df.isnull().sum() / len(imputed_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object            0                           0.0
company_name            object         2596                          23.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object         1502                          14.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

**Impute data for company_name column**

In [18]:
# Impute 'sector' as 'missing' for rows with null sector values
imputed_df['company_name'].fillna('missing', inplace=True)

In [19]:
# Get the data types by column
data_types_by_column = imputed_df.dtypes

# Get the number of null values by column
null_values_by_column = imputed_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((imputed_df.isnull().sum() / len(imputed_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object            0                           0.0
company_name            object            0                           0.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object         1502                          14.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

**Impute data for special_instructions column**

In [20]:
# Impute 'sector' as 'missing' for rows with null sector values
imputed_df['special_instructions'].fillna('missing', inplace=True)

In [21]:
# Get the data types by column
data_types_by_column = imputed_df.dtypes

# Get the number of null values by column
null_values_by_column = imputed_df.isnull().sum()

# Get the percentage of null values by column
percentage_of_null_values = round((imputed_df.isnull().sum() / len(imputed_df)) * 100)

# Combine the data types and null values into a DataFrame
data_info = pd.DataFrame({'Data Type': data_types_by_column, 'Null Values': null_values_by_column, 'Percentage of null values(%)': percentage_of_null_values})

# Print the result
data_info

Data Type  Null Values  Percentage of null values(%)
event_id                object            0                           0.0
event_state             object            0                           0.0
sales_type              object            0                           0.0
sector                  object            0                           0.0
company_name            object            0                           0.0
customer_id              int64            0                           0.0
cost_per_pax           float64            0                           0.0
pax                      int64            0                           0.0
address                 object            0                           0.0
postal_code              int64            0                           0.0
special_instructions    object            0                           0.0
pickup_date_time        object            0                           0.0
serving_date_time       object            0                           0.0
revenue                float64            0                           0.0

Now that all the data has been imputted, we wil sort the dataframe by date and time.

In [22]:
# Sort the dataframe by time
imputed_df = imputed_df.sort_values(by='pickup_date_time', ascending=True)
imputed_df.reset_index(drop=True, inplace=True)
imputed_df.head()

event_id event_state sales_type                                 sector  \
0    F3253   PUBLISHED    inbound                  Technology (software)   
1    F3201   PUBLISHED     repeat                         Pharmaceutical   
2    F3259   PUBLISHED    inbound  Government, Nonprofits & Associations   
3    F3239   PUBLISHED     repeat                              Hospitals   
4    F3254   PUBLISHED    inbound                  Technology (software)   

                            company_name  customer_id  cost_per_pax  pax  \
0  Amazon Web Services Singapore Pte Ltd        50296         16.73   23   
1           Novartis (Singapore) Pte Ltd        26299          7.43   30   
2                   Ministry of Manpower        50306         16.73   39   
3     KK Women's and Children's Hospital        47121          9.30   50   
4  Amazon Web Services Singapore Pte Ltd        50296          0.00   23   

                                             address  postal_code  \
0     8 Marina View, #23-01 Asia Square Tower 1             18960   
1  1E Kent Ridge Rd, National University Cancer I...       119228   
2  18 Havlock Road, Main Entrance, Ministry of Ma...        59764   
3  100 Bukit Timah Road, Department of Maternal F...       229899   
4     8 Marina View, #23-01 Asia Square Tower 1             18960   

                                special_instructions     pickup_date_time  \
0            call Mona upon upon arrival at level 23  2022-01-06T07:00:00   
1  30 x Bento + Cutlery | 39 x ripple cups, lids,...  2022-01-06T07:15:00   
2                                            missing  2022-01-06T10:30:00   
3  50 x bento + cutlery | \n\nMs. Jasmine Lim – 9...  2022-01-06T10:30:00   
4                                            missing  2022-01-06T10:30:00   

     serving_date_time  revenue  
0  2022-01-06T08:00:00   399.79  
1  2022-01-06T08:15:00   308.13  
2  2022-01-06T11:30:00   661.77  
3  2022-01-06T11:30:00   465.00  
4  2022-01-06T11:30:00    15.00

**Get the new postal code to have leading zeros**

Some of the postal codes have 5 digits which is not correct as there is supposed to be a 0 in front. Therefore, we will add the leading zeros to the postal codes.

In [26]:
# Assuming 'df' is your original DataFrame
new_df1 = imputed_df.copy()

# Convert the 'postal_code' column to integer
new_df1['postal_code'] = new_df1['postal_code'].astype(str)

# Define a function to add a leading zero to 5-digit postal codes
def add_leading_zero(p_code):
    if len(p_code) == 5:
        return '0' + p_code
    else:
        return p_code

# Apply the function to the 'postal_code' column in the DataFrame
new_df1['postal_code_new'] = new_df1['postal_code'].apply(add_leading_zero)

# Convert the 'postal_code' column back to integer
new_df1['postal_code'] = new_df1['postal_code'].astype(int)

# Display the sorted DataFrame
new_df1.head()

event_id event_state sales_type                                 sector  \
0    F3253   PUBLISHED    inbound                  Technology (software)   
1    F3201   PUBLISHED     repeat                         Pharmaceutical   
2    F3259   PUBLISHED    inbound  Government, Nonprofits & Associations   
3    F3239   PUBLISHED     repeat                              Hospitals   
4    F3254   PUBLISHED    inbound                  Technology (software)   

                            company_name  customer_id  cost_per_pax  pax  \
0  Amazon Web Services Singapore Pte Ltd        50296         16.73   23   
1           Novartis (Singapore) Pte Ltd        26299          7.43   30   
2                   Ministry of Manpower        50306         16.73   39   
3     KK Women's and Children's Hospital        47121          9.30   50   
4  Amazon Web Services Singapore Pte Ltd        50296          0.00   23   

                                             address  postal_code  \
0     8 Marina View, #23-01 Asia Square Tower 1             18960   
1  1E Kent Ridge Rd, National University Cancer I...       119228   
2  18 Havlock Road, Main Entrance, Ministry of Ma...        59764   
3  100 Bukit Timah Road, Department of Maternal F...       229899   
4     8 Marina View, #23-01 Asia Square Tower 1             18960   

                                special_instructions     pickup_date_time  \
0            call Mona upon upon arrival at level 23  2022-01-06T07:00:00   
1  30 x Bento + Cutlery | 39 x ripple cups, lids,...  2022-01-06T07:15:00   
2                                            missing  2022-01-06T10:30:00   
3  50 x bento + cutlery | \n\nMs. Jasmine Lim – 9...  2022-01-06T10:30:00   
4                                            missing  2022-01-06T10:30:00   

     serving_date_time  revenue postal_code_new  
0  2022-01-06T08:00:00   399.79          018960  
1  2022-01-06T08:15:00   308.13          119228  
2  2022-01-06T11:30:00   661.77          059764  
3  2022-01-06T11:30:00   465.00          229899  
4  2022-01-06T11:30:00    15.00          018960

With the imputted data and sorted dataframe, we will save the file as a new csv file and move on to exploratory data analysis.

In [27]:
new_df1.to_csv('../data/imputed_data.csv', index=False)